rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -218,9 +248,35 @@ std::string ZstdCompressor::compress_uri(const std::string & uri)
 }
 
 void ZstdCompressor::compress_serialized_bag_message(
-  rosbag2_storage::SerializedBagMessage *)
+  rosbag2_storage::SerializedBagMessage * message)
 {
-  throw std::logic_error{"Not implemented"};
+  const auto start = std::chrono::high_resolution_clock::now();
+  // Allocate based on compression bound and compress
+  const auto uncompressed_buffer_length =
+    ZSTD_compressBound(message->serialized_data->buffer_length);
+  std::vector<uint8_t> compressed_buffer(uncompressed_buffer_length);
+
+  // Perform compression and check.
+  // compression_result is either the actual compressed size or an error code.
+  const auto compression_result = ZSTD_compress(
+    compressed_buffer.data(), compressed_buffer.size(),
+    message->serialized_data->buffer, message->serialized_data->buffer_length,
+    kDefaultZstdCompressionLevel);
+  throw_on_zstd_error(compression_result);
+
+  // Compression_buffer_length might be larger than the actual compression size
+  // Resize compressed_buffer so its size is the actual compression size.
+  compressed_buffer.resize(compression_result);
+
+  const auto resize_result =
+    rcutils_uint8_array_resize(message->serialized_data.get(), compression_result);
```

Comment:
Not familiar with the internals of `rcutils_uint8_array_resize` but does it handle resizing the `buffer` and updating the `buffer_length`?

If so, why perform the resize here when the next operation already performs that? (275-276)

Reply:
To be honest, I am also not familiar with the internals of `rcutils_uint8_array_resize` and haven't yet found very clear examples of how it's supposed to work.  It appears that after calling it, it does resize how much memory is allocated, but it does not actually change the value of `buffer_length`.  This code as it is right now seems to work, although I still need to dig deeper into it to figure out whether this is the intended usage or not.

Reply:
Looking at how this function is used in https://github.com/ros2/rcutils/blob/master/test/test_uint8_array.cpp#L52, I believe that it changes the buffer's capacity but not its length, so manually setting the length here is normal.